File: leetcode/medium/HandOfStraights.java
import java.util.TreeMap;

/**
 * Solution for the Hand of Straights problem.
 */
class Solution {
    /**
     * Checks if a hand of cards can be rearranged into groups of `groupSize`
     * consecutive cards.
     *
     * The approach is greedy. We first count the occurrences of each card using a
     * sorted map (TreeMap). A sorted map is crucial because it allows us to always
     * start a new potential straight with the smallest card currently available.
     *
     * The algorithm proceeds as follows:
     * 1. Check if the total number of cards is divisible by `groupSize`. If not,
     *    it's impossible, so return false.
     * 2. Count all cards and store them in a TreeMap to keep them sorted by card value.
     * 3. As long as there are cards left in the map:
     *    a. Get the smallest card number available (the first key in the TreeMap).
     *       This will be the start of a new straight.
     *    b. Iterate from this start card up to `startCard + groupSize - 1`.
     *    c. For each card in this required sequence, decrement its count in the map.
     *    d. If a required card is not found in the map, then a straight cannot be
     *       formed, and we return false.
     * 4. If we successfully form all groups, return true.
     *
     * @param hand An array of integers representing the cards.
     * @param groupSize The size of each group of consecutive cards.
     * @return True if the hand can be rearranged as required, false otherwise.
     */
    public boolean isNStraightHand(int[] hand, int groupSize) {
        // 1. Initial check for divisibility.
        if (hand.length % groupSize != 0) {
            return false;
        }

        // 2. Count card occurrences in a sorted map.
        TreeMap<Integer, Integer> cardCounts = new TreeMap<>();
        for (int card : hand) {
            cardCounts.put(card, cardCounts.getOrDefault(card, 0) + 1);
        }

        // 3. Greedily form straights until all cards are used.
        while (!cardCounts.isEmpty()) {
            // a. Get the smallest card to start a new straight.
            int startCard = cardCounts.firstKey();

            // b. Try to form a straight of `groupSize` cards.
            for (int i = 0; i < groupSize; i++) {
                int currentCard = startCard + i;

                // c. Check if the required card is available.
                if (!cardCounts.containsKey(currentCard)) {
                    return false;
                }

                // d. Decrement the card's count.
                int count = cardCounts.get(currentCard);
                if (count == 1) {
                    // Remove the card from the map if it's the last one.
                    cardCounts.remove(currentCard);
                } else {
                    cardCounts.put(currentCard, count - 1);
                }
            }
        }

        // 4. If the loop completes, all cards have been formed into valid straights.
        return true;
    }
}

File: leetcode/medium/HandOfStraights.py
from collections import Counter
import heapq
from typing import List

class Solution:
    """
    Determines if a hand of cards can be partitioned into consecutive groups of a specified size.
    This is commonly known as the "Hand of Straights" problem (or "Divide Array in Sets of K Consecutive Numbers").
    """
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        """
        :param hand: A list of card integer values.
        :param groupSize: The required size of each consecutive group (straight).
        :return: True if the partitioning is possible, False otherwise.
        """
        n = len(hand)
        
        # 1. Initial Check: The total number of cards must be divisible by the group size.
        if n % groupSize != 0:
            return False
            
        # 2. Count Card Frequencies: Store the frequency of each card value. O(N)
        # Using a hash map (Counter) for O(1) average time lookups.
        counts = Counter(hand)
        
        # 3. Min-Heap Initialization: Store the unique card values in a min-heap. O(U log U)
        # U is the number of unique cards. This allows efficient access to the smallest available card.
        min_heap = list(counts.keys())
        heapq.heapify(min_heap)
        
        # 4. Greedy Processing
        # The greedy strategy dictates that the smallest available card must start a new group.
        while min_heap:
            # Get the smallest available card value to start a new straight.
            start_card = min_heap[0]
            
            # If the smallest card in the heap has already been fully used (its count is 0), 
            # pop it and move to the next smallest card. This ensures min_heap[0] is always relevant.
            if counts[start_card] == 0:
                heapq.heappop(min_heap)
                continue # Skip to the next iteration
            
            # If the start_card count is > 0, try to form a straight of size `groupSize`.
            for card in range(start_card, start_card + groupSize):
                # a. Check if the required consecutive card is available. O(1) average.
                if counts[card] == 0:
                    # If any card in the required straight is missing, the partition is impossible.
                    return False
                
                # b. Use one instance of the current card.
                counts[card] -= 1
                
            # OPTIMIZATION/CORRECTION:
            # We ONLY check and pop the `start_card` from the heap after successfully forming 
            # the straight. The only card guaranteed to be removed from the heap (if its count hits 0) 
            # *must* be the current smallest element (`start_card`). The loop's starting check 
            # handles the removal for all other consumed cards.
            # No explicit `heappop` is needed inside the inner loop for `card` != `start_card`.
            # If the count for `start_card` is now 0, it means the group that consumed its last 
            # instance was formed, so we can now safely pop it.
            if counts[start_card] == 0:
                 heapq.heappop(min_heap)

        # 5. Result
        # If the loop completes, all cards have been successfully grouped.
        return True